src/bin.ts

```diff
@@ -403,23 +407,76 @@ function phase3(payload: BootstrapState) {
       experimentalSpecifierResolution as ExperimentalSpecifierResolution,
   });
 
-  if (preloadedConfig.options.esm) payload.shouldUseChildProcess = true;
+  // If ESM is enabled through the parsed tsconfig, stage4 should be run in a child
+  // process with the ESM loaders configured.
+  if (preloadedConfig.options.esm) {
+    payload.shouldUseChildProcess = true;
+  }
+
   return { preloadedConfig };
 }
 
+/**
+ * Determines the entry-point information from the argv and phase2 result. This
+ * method will be invoked in two places:
+ *
+ *   1. In phase 3 to be able to find a project from the potential entry-point script.
+ *   2. In phase 4 to determine the actual entry-point script.
+ *
+ * Note that we need to explicitly re-resolve the entry-point information in the final
+ * stage because the previous stage information could be modified when the bootstrap
+ * invocation transitioned into a child process for ESM.
+ *
+ * Stages before (phase 4) can and will be cached by the child process through the Brotli
+ * configuration and entry-point information is only reliable in the final phase. More
+ * details can be found in here: https://github.com/TypeStrong/ts-node/issues/1812.
+ */
```

Comment:
Thanks for adding all these comments.

It makes sense to me that we must repeat resolution in phase 4.  Are there any situations where we know, without any risk, that we can skip this duplicate resolution?  It would be a performance optimization.

If we cannot safely do this, then we shouldn't.  But if there's a safe way to do this, then I might try my hand at adding it.

Reply:
I think we could re-use this when the `argvResult` has not changed. We only need the rerun this function due to the patches in the child loader entry-point. We could indicate this change through a flag and request re-computation.

This feels like a micro optimization through, not worth it personally as this is just some simple logical expressions.